pilot/pkg/model/push_context.go

```diff
@@ -310,15 +312,50 @@ func (ps *PushContext) UpdateMetrics() {
 }
 
 // Services returns the list of services that are visible to a Proxy in a given config namespace
-func (ps *PushContext) Services(proxy *Proxy) []*Service {
+func (ps *PushContext) Services(p *Proxy) []*Service {
 	// TODO: use network scopes here, and return services in the proxy namespace if
 	// configured to use current_namespace as the scope and proxy is not nil
-	//if proxy == nil {
-	//	return ps.allServices
-	//}
+	if p == nil {
+		return ps.allServices
+	}
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+	if p.serviceDependencies != nil {
+		return p.serviceDependencies
+	}
 	return ps.allServices
 }
 
+// OnConnect is called when a node connects. Will update per-node data.
+func (ps *PushContext) OnConnect(p *Proxy) {
+	// For now Router (Gateway) is not using the isolation - the Gateway already has explicit
+	// bindings.
+	if pilot.NetworkScopes != "" && p.Type == Sidecar {
```

Comment:
why aren't you using the stuff that was added to the global mesh config for this purpose?